### ISSUE TITLE
Optionally download files via HTTP

### ISSUE THREAD

**Issue by coopernurse** [dogestry/dogestry#68]
Opening this to start a discussion.  Feel free to close if you feel it's not appropriate for the main project.

I have a client that runs their own hardware (i.e. external to AWS), so bandwidth to AWS is a consideration.  In the production zone they have over a dozen VMs in a cluster.

When releases are done, all VMs pull a number of images and restart them. This causes considerable bandwidth contention.

I realize dogestry has the `-pullhosts` flag, but if I understand it correctly, hosts must expose the docker daemon on a TCP port, which is not something we want to do for security reasons.

The approach we settled on was to run a nginx proxy with the nginx_aws_auth module compiled in, configured as a caching reverse proxy to our private S3 bucket.  

The only change to dogestry was to `S3Remote.getFile()` so that it optionally downloads files from the HTTP endpoint, rather than from S3 directly.  

I made this patch on my fork here:

https://github.com/coopernurse/dogestry/commit/db5603f57d23b4d02c473b8962738802885f8462

We run nginx as a docker container using this image:

https://registry.hub.docker.com/u/coopernurse/nginx-s3-proxy/

If we think this is a generally useful addition I'd be happy to clean up this implementation a bit and try to land it in the main repo.  It may simply be that I misunderstand the security implications of `-pullhosts`, which appears to be the supported way to solve the concurrent pull problem.

Thanks!

**Comment by didip** [comment]
Interesting... In a separate conversation, we have talked about running dogestry as HTTP service.

If dogestry is run as a service, then it can expose proxy endpoints to S3 backend. This way, you don't need nginx anymore. It's just dogestry talking to another dogestry. The authentication story can be more flexible too.

What do you think?

**Comment by coopernurse** [comment]
yeah, that could be good although I wonder if we'd just be re-implementing a bunch of nginx's caching implementation.  folks will want to dial in things like cache size, TTL, etc.

Docker should have used the BitTorrent DHT right?  I'm only half joking. I wonder if anyone has tried to do that.

**Comment by didip** [comment]
Ha! funny you should mentioned it. I talked to the guy who wrote this: https://github.com/mafintosh/torrent-docker

**Comment by didip** [comment]
That said, about Go HTTP proxy, a lot of work have gone through https://github.com/mailgun/oxy & https://github.com/mailgun/vulcand. If I were to write an http proxy, I would start with oxy as library.

Your current idea, with a little clean up, is more or less fine. I am just wondering about how general purpose it is.

For example, should we worry about HTTP basic auth? What about S3 auth headers? Or should we just expect that users deal with auth on their own just like your use case?

**Comment by dselans** [comment]
@coopernurse does the most recent addition of dogestry server mode address one of your issues?

**Comment by coopernurse** [comment]
@dselans I haven't tried the server mode.  The nginx-proxy setup has been working fine for us.  If we were starting from scratch I suspect this might work, although it looks like we'd have to run an extra daemon on each host.  I'm also a little thoughtful about security - presumably anyone with access to that port can push arbitrary images to your docker daemon - is that right?
